src/effect.cpp

```diff
@@ -2,11 +2,168 @@
+static std::string const eff_arg_tick       = {"tick"};
+
+static std::string const eff_type_str     {"STR"};
+static std::string const eff_type_dex   = {"DEX"};
+static std::string const eff_type_per   = {"PER"};
+static std::string const eff_type_int   = {"INT"};
+static std::string const eff_type_speed = {"SPEED"};
+
+static std::string const eff_type_pain    = {"PAIN"};
+static std::string const eff_type_hurt    = {"HURT"};
+static std::string const eff_type_sleep   = {"SLEEP"};
+static std::string const eff_type_pkill   = {"PKILL"};
+static std::string const eff_type_stim    = {"STIM"};
+static std::string const eff_type_health  = {"HEALTH"};
+static std::string const eff_type_h_mod   = {"H_MOD"};
+static std::string const eff_type_rad     = {"RAD"};
+static std::string const eff_type_hunger  = {"HUNGER"};
+static std::string const eff_type_thirst  = {"THIRST"};
+static std::string const eff_type_fatigue = {"FATIGUE"};
+static std::string const eff_type_cough   = {"COUGH"};
+static std::string const eff_type_vomit   = {"VOMIT"};
+
+/**
+ * clamp n to [lo, hi]
+ */
+template <typename T>
+inline T clamp(T const n, T const lo, T const hi) noexcept {
+    return (n < lo) ? lo :
+           (n > hi) ? hi :
+           n;
+}
+
+
+// This should probably be moved elsewhere if it is to be kept so that it can be more widely used
+namespace detail {
+
+template <typename To, typename From>
+struct select_result_type {
+    using type = typename std::conditional<
+        std::is_same<void, To>::value, From, To
+    >::type;
+};
+
+template <typename To, typename From>
+using select_result_type_t = typename select_result_type<To, From>::type;
+
+} //namespace detail
+
+/**
+ * do a round followed by a cast
+ *
+ * @tparam To optional cast type, the results will otherwise fallback to decltype(value)
+ */
+template <
+    typename To = void
+  , typename From
+  , typename Result = detail::select_result_type_t<To, From>
+  , typename std::enable_if<std::is_floating_point<From>::value>::type* = nullptr
+>
+auto round_to(From const value) -> Result {
+    using std::round;
+    return static_cast<To>(round(value));
+}
+
+/**
+ * mapping from hash value to T
+ */
+template <typename T>
+using hash_pair = std::pair<int, T>;
+
+/**
+ * find key in an array
+ *
+ * @return a pair of the form {value at key, key was found}
+ */
+template <typename T, typename K, size_t N>
+std::pair<T, bool>
+find_mapping(hash_pair<T> const (&arr)[N], K const& key) {
+    auto const beg = std::begin(arr);
+    auto const end = std::end(arr);
+
+    auto const it = std::find_if(beg, end, [&](hash_pair<T> const& value) {
+        return key == value.first;
+    });
+    
+    auto const found = (it != end);
+
+    return std::make_pair(it->second, found);
+}
+
+/**
+ * get a value (enumeration) from a string value
+ *
+ * @return a pair of the form {value, string was valid}
+ */
+template <typename T>
+std::pair<T, bool> from_string(std::string const& s);
```

Comment:
While I agree that having a simple string <-> enum mapping is handy, this is far from simple, any particular reason you don't just make a lookup table and use it inline?  All the templating and automatic dispatch here just serves to make it less readable rather than more.